ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala

```diff
@@ -44,31 +41,56 @@ object KeyValueConsumption {
 
     entry.getPayloadCase match {
       case DamlLogEntry.PayloadCase.PACKAGE_UPLOAD_ENTRY =>
-        entry.getPackageUploadEntry.getArchivesList.asScala.map { archive =>
+        val participantId =
+          Ref.LedgerString.assertFromString(entry.getPackageUploadEntry.getParticipantId)
+        //only send package uploaded update if the log entry includes an archive
+        val uniquePackages = entry.getPackageUploadEntry.getArchivesList.asScala.map { archive =>
           Update.PublicPackageUploaded(
             archive,
             if (entry.getPackageUploadEntry.getSourceDescription.nonEmpty)
               Some(entry.getPackageUploadEntry.getSourceDescription)
             else None,
-            Ref.LedgerString.assertFromString(entry.getPackageUploadEntry.getParticipantId),
-            recordTime
+            participantId,
+            recordTime,
+            Ref.LedgerString.assertFromString(entry.getPackageUploadEntry.getSubmissionId)
           )
-        }(breakOut)
+        }.toList
+        uniquePackages ++ List(
+          Update
+            .PackageUploadEntryAccepted(participantId, entry.getPackageUploadEntry.getSubmissionId))
 
       case DamlLogEntry.PayloadCase.PACKAGE_UPLOAD_REJECTION_ENTRY =>
-        List.empty
+        val rejection = entry.getPackageUploadRejectionEntry
+        val participantId = rejection.getParticipantId
+        val proposedPackageUpload = rejection.getInvalidPackage
+
+        List(
+          Update.PackageUploadEntryRejected(
+            Ref.LedgerString.assertFromString(participantId),
+            Ref.LedgerString.assertFromString(rejection.getSubmissionId),
+            reason = rejection.getReasonCase match {
+              case DamlPackageUploadRejectionEntry.ReasonCase.INVALID_PACKAGE =>
+                s"Package ${proposedPackageUpload.getDetails} rejected as invalid"
+              case DamlPackageUploadRejectionEntry.ReasonCase.REASON_NOT_SET =>
+                s"Package ${proposedPackageUpload.getDetails} upload failed for undetermined reason"
+              case DamlPackageUploadRejectionEntry.ReasonCase.PARTICIPANT_NOT_AUTHORIZED =>
+                s"Participant ${participantId} not authorized to upload package"
```

Comment:
You don't need `{}` for interpolating a variable reference.

Reply:
resolved